### DISCUSSION TITLE
Localization in packages

### DISCUSSION THREAD

**Discussion by Fexonchik5120** [termux/termux-packages#18297] [Q&A] [answered]
Is it possible to compile packages with localization support? Or is this not possible due to the limitations of the Termux application?

**Comment by sylirre** [comment, marked as answer]
Termux environment itself has very limited support for locales, mostly because it uses Android libc (Bionic) rather than GNU libc which has full-fledged for them.

We have some stubs providing functionality necessary to compile locale-depending programs without issues:

* https://github.com/termux/termux-packages/blob/62f2443fcdf2acad4af48e748cb3f2017d40276f/ndk-patches/langinfo.h
* https://github.com/termux/termux-packages/blob/62f2443fcdf2acad4af48e748cb3f2017d40276f/ndk-patches/libintl.h

These of course do not add multi-language support.

Another reason that Termux app and all docs (e.g. Termux Wiki) are English-only and we can't provide translations with proper quality.

Localization task has the lowest priority for us, if we'll ever would work on it.

If someone wants to solve this, *it would be required to work on both parts of Termux (application + package) as well as maintaining localization further to keep everything in sync*.

**Reply by Fexonchik5120** [reply]
OK, I understand, but then why is the gettext package in the Termux repository? And is it possible to implement localization as in Links? (gettext is used there, but localization is configured in the app itself, it also work in termux)

**Reply by sylirre** [reply]
`gettext` provided solely as dependency for some packages which can't be compiled without it. Also I already told you that underlying locales functionality is either not supported or is broken.

Gettext isn't used standalone. Without proper locales support it is basically a no-op.

The most recent Android libc provides support only for these locales:
```
static bool __is_supported_locale(const char* locale_name) {
  return (strcmp(locale_name, "") == 0 ||
          strcmp(locale_name, "C") == 0 ||
          strcmp(locale_name, "C.UTF-8") == 0 ||
          strcmp(locale_name, "en_US.UTF-8") == 0 ||
          strcmp(locale_name, "POSIX") == 0);
}
```

Source: https://android.googlesource.com/platform/bionic/+/refs/heads/main/libc/bionic/locale.cpp (commit `4cc96c1d1eaef69c7f5a47656f6d3cbd65054567`).

This is *hardcoded*. We obviously choose `en_US.UTF-8` and that's fixed configuration.

Application localization has nothing to do with localization of packages. Config of one doesn't affect another. Android implements locales in entirely different way than standard Linux distro. The code shown above clearly tells that low-level Android part (dependency of Termux packages) supports only bare minimum.

**Comment by sukronc** [comment]
> Termux environment itself has very limited support for locales, mostly because it uses Android libc (Bionic) rather than GNU libc which has full-fledged for them.
> 
> We have some stubs providing functionality necessary to compile locale-depending programs without issues:
> 
> * https://github.com/termux/termux-packages/blob/62f2443fcdf2acad4af48e748cb3f2017d40276f/ndk-patches/langinfo.h
> * https://github.com/termux/termux-packages/blob/62f2443fcdf2acad4af48e748cb3f2017d40276f/ndk-patches/libintl.h
> 
> These of course do not add multi-language support.
> 
> Another reason that Termux app and all docs (e.g. Termux Wiki) are English-only and we can't provide translations with proper quality.
> 
> Localization task has the lowest priority for us, if we'll ever would work on it.
> 
> If someone wants to solve this, *it would be required to work on both parts of Termux (application + package) as well as maintaining localization further to keep everything in sync*.
